### PR TITLE
Tidy and document the codebase.

### LINKED ISSUE #177: There are hundreds of "NametagBackgroundMaterial" in the material list when viewed with the inspector.
This COULD be contributing to slowdowns as the inspector appears to shows hundreds of materials being used, despite just a handful in the scene. Needs further investigation on a locally run client.

0ff6e2952293e02f5dc0d45afe0546b9 1

Worth noting there quite a few "skybox" materials appearing as well.

### PR DESCRIPTION
**Pull request by Gigabyte5671**
This PR adds 141 documenting comments, removes 23 uses of `eslint-disable`, and addresses a few bugs.

Bugs Addressed:
- Fix #177 with [cd3aad4](https://github.com/vircadia/vircadia-web/pull/179/commits/cd3aad4da387fd65ab975c6ed38945460c604ad7).
- The first-time-setup wizard could skip the world-selection step even when no location was pending.
- Various loosely-typed conditionals preventing number values from being set to `0`. [699bdaf](https://github.com/vircadia/vircadia-web/pull/179/commits/699bdafeac99baf8321b7bca7b7dbcd2440566e0).

Features added:
- An FPS counter in the bottom right corner of the window, toggleable from the graphics settings menu.
- The history of all log messages can be dumped to a text file via a button in the help menu.

To be addressed in further PRs:
- (AudioIO) The function for calculating the audio level from the user's microphone should be rewritten as an AudioWorkletProcessor instead of the deprecated ScriptProcessorNode.
- (`object/GameObject.ts:165`) The method of removing a GameObject from the scene is rather peculiar. This should be documented with comments and tested.
- (entity/components/components/skybox.ts) Any changes to the skybox colour (from the Domain server) make it go transparent for a few seconds.

### CODE REVIEW

File: src/components/overlays/jitsi/Jitsi.vue [hunk 1]
```diff
@@ -59,20 +59,24 @@ export default defineComponent({
...
-    }),
-
     components: {
         OverlayShell
     },
+
+    data() {
+        return {
+            room: <JitsiRoomInfo><unknown>undefined,
+            jitisElement: <HTMLElement><unknown>undefined
```
Comment (ctrlaltdavid) [inline src/components/overlays/jitsi/Jitsi.vue:69]:
```suggestion
            jitsiElement: <HTMLElement><unknown>undefined
```
Here and elsewhere?

File: src/modules/avatar/controller/inputs/keyboardInput.ts [hunk 2]
```diff
@@ -127,31 +129,39 @@ export class KeyboardInput implements IInputHandler {
...
+    private _onKeyDown(event: ActionEvent): void {
+        const sourceEvent = event.sourceEvent as Event;
+        if (!sourceEvent || !(sourceEvent instanceof KeyboardEvent) || !sourceEvent.code) {
             return;
         }
 
-        this._inputMap[evt.sourceEvent.code] = evt.sourceEvent.type === "keydown";
-        this._shiftKey = evt.sourceEvent.shiftKey === true;
+        this._inputMap[sourceEvent.code] = sourceEvent.type === "keydown";
+        this._shiftKey = sourceEvent.shiftKey === true;
```
Comment (ctrlaltdavid) [inline src/modules/avatar/controller/inputs/keyboardInput.ts:139]: `=== true` here and elsewhere looks suspicious: if the variable on the left is a `boolean` then don't need to test it against `true`.

File: src/modules/debugging/log.ts [hunk 3]
```diff
@@ -2,94 +2,162 @@
...
+        this._logHistory.push([time, type, level, message, ...optionalMessages].join(" "));
     }
 
-    // debug log message. Output if level is set to DEBUG
-    function debug(pType: types, pMsg: string) {
-        if (logLevel === levels.DEBUG) {
-            print(pType, levels.DEBUG, pMsg);
+    /**
+     * Print a debug message to the log.
+     * (Ignored if the log level is not set to `DEBUG`).
```
Comment (ctrlaltdavid) [inline src/modules/debugging/log.ts:97]:
```suggestion
     * (Ignored if the log level is not set to `DEBUG`.)
```

File: src/modules/debugging/log.ts [hunk 4]
```diff
@@ -2,94 +2,162 @@
...
+        this.print(type, LogLevel.ERROR, message, ...optionalMessages);
     }
 
-    // warn log message. output if level is warn or above
-    function warn(pType: types, pMsg: string) {
-        if (logLevel in [levels.WARN, levels.DEBUG, levels.ERROR]) {
-            print(pType, levels.WARN, pMsg);
+    /**
+     * Print a warning message to the log.
+     * (Ignored if the log level is not set to `WARN`, `DEBUG`, or `ERROR`).
```
Comment (ctrlaltdavid) [inline src/modules/debugging/log.ts:118]:
```suggestion
     * (Ignored if the log level is not set to `WARN`, `DEBUG`, or `ERROR`.)
```

File: src/modules/domain/client.ts [hunk 5]
```diff
@@ -1,47 +1,52 @@
...
     DISCONNECTED,
     CONNECTED
 }
-/** Parant class for all AssignmentClients  */
-export abstract class Client {
 
-    /** Return string name of AssignmentClientState.
-     * This is needed since the SDK does not export AssignmentClient
+/**
+ * Parant class for all AssignmentClients.
```
Comment (ctrlaltdavid) [inline src/modules/domain/client.ts:22]:
```suggestion
 * Parent class for all AssignmentClients.
```

File: src/modules/domain/client.ts [hunk 6]
```diff
@@ -56,16 +61,16 @@ export abstract class Client {
      * A better implmentation would be to save the (resolve,reject) in a list and
      * call them when the onStateChanged event happens.
      *
-     * @param pTimeoutMS optional number of MS to wait. Default is to 5 minutes.
-     * @returns the Client being waited on
-     * @throws exception if waited more than the timeout interval
+     * @param timeoutMS `(Optional)` The number of milliseconds to wait. Default is `300,000` (5 minutes).
```
Comment (ctrlaltdavid) [inline src/modules/domain/client.ts:64]:
```suggestion
     * @param [timeoutMS] The number of milliseconds to wait. Default is `300,000` (5 minutes).
```
Using square brackets should work for indicating optional parameters.

**Review by ctrlaltdavid — APPROVED** [review]
A few minor things noticed in passing to address either in this PR or subsequently.

Note also that the vircadia-web-sdk.d.ts file should not be necessary - the typings are included in the Web SDK.

File: src/modules/domain/client.ts (hunk 6 above)
Comment (Gigabyte5671) [reply src/modules/domain/client.ts:64]: I've tried:
- @param [timeoutMS]
- @param [timeoutMS=]
- @param [timeoutMS=300000]
- @param {number} [timeoutMS]
- @param {number=} [timeoutMS]
- @param {number|undefined} [timeoutMS]

Unfortunately, none of these appear to mark the parameter as optional. All produce the same output:
image

Is there a particular setting I should change for these to work?

Comment (ctrlaltdavid) [reply src/modules/domain/client.ts:64]: Hmmm, OK. Square brackets should work for TypeDoc / JSDoc, but I don't know about Intellisense conventions.
Something to keep a look out for in our travels.

Comment (Gigabyte5671) [reply src/modules/domain/client.ts:64]: Roger, I'll keep an eye out.

**Pull request merged**